### PR TITLE
Mamba Implementation

### PR DESCRIPTION
**Pull request by reddyn12**
Redo - #3251 

Got advice from people in the discord and changed the pull request, but never updated the description. I also added tests for the two smaller models.

### CODE REVIEW

File: test/models/test_mamba.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@unittest.skipIf(CI and Device.DEFAULT in ["CLANG"], "Inference hack broken on CLANG")
+class TestMamba(unittest.TestCase):
+  def test_mamba_130M(self):
+    OUT_130M = '''Why is gravity \nnot a good idea?\n\nA:'''
+    model = Mamba.from_pretrained('130m')
+    tinyoutput = generate(model, TOKENIZER, PROMPT, n_tokens_to_gen=10)
+    self.assertEqual(OUT_130M, tinyoutput)
+    del model
+  def test_mamba_370M(self):
+    OUT_370M = '''Why is gravity \nso important?\nBecause it's the only'''
```
Comment (geohot) [inline test/models/test_mamba.py:19]: How did you verify this is correct?

File: test/models/test_mamba.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import unittest
+from tinygrad.helpers import CI
+from tinygrad import Device
+from examples.mamba import Mamba, generate
+from transformers import AutoTokenizer
+
+PROMPT = 'Why is gravity '
+TOKENIZER = AutoTokenizer.from_pretrained("EleutherAI/gpt-neox-20b")
+
+@unittest.skipIf(CI and Device.DEFAULT in ["CLANG"], "Inference hack broken on CLANG")
```
Comment (geohot) [inline test/models/test_mamba.py:10]: What's broken on CLANG and why?

File: test/models/test_mamba.py (hunk 1 above)
Comment (reddyn12) [reply test/models/test_mamba.py:19]: Ran it on a troch implementation: https://github.com/reddyn12/tinygrad/blob/e9dfa6fe57e689df4e5d342989cd236bf317f407/examples/compare/mamba/mambaTorch.py

Comment (geohot) [reply test/models/test_mamba.py:19]: Why is it putting in the \n? I have never seen an LLM do that

Comment (reddyn12) [reply test/models/test_mamba.py:19]: How else can a new line happen?

Comment (reddyn12) [reply test/models/test_mamba.py:19]: I can try with pure "enter" but shouldn't be a difference right?

Comment (geohot) [reply test/models/test_mamba.py:19]: That's not what I mean. Why is `\n` the next symbol?

Comment (reddyn12) [reply test/models/test_mamba.py:19]: Oh, maybe a training thing? poem roleplay?

File: test/models/test_mamba.py (hunk 2 above)
Comment (reddyn12) [reply test/models/test_mamba.py:10]: After some digging, it's not what I thought.  
(logits[:, -1, :]).softmax().argmax(axis=-1).item() ALWAYS returns 50279 which is an empty string
This make me feel like softmax may be an issue cuz the probability of next token being empty should be the least, right?

Comment (reddyn12) [reply test/models/test_mamba.py:10]: After some testing on a Tensor/torch.arange(16).reshape(4,4):
.softmax works properly
.argmax works properly
.softmax.argmax gets incorrect result ONLY FOR CLANG=1, works on CUDA=1

Comment (dagelf) [reply test/models/test_mamba.py]: That's not a good excuse to skip a test.

Comment (reddyn12) [reply test/models/test_mamba.py]: Issue #3552

File: test/models/test_mamba.py (hunk 1 above)
Comment (geohot) [reply test/models/test_mamba.py:19]: I simply don't believe this. Does mamba on huggingface do this?

Comment (reddyn12) [reply test/models/test_mamba.py:19]: Ya:
```python 
from transformers import AutoModelForCausalLM , AutoTokenizer

model = AutoModelForCausalLM.from_pretrained('Q-bert/Mamba-130M', trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained('Q-bert/Mamba-130M')

text = 'Why is gravity '

input_ids = tokenizer.encode(text, return_tensors="pt")

output = model.generate(input_ids, max_length=20, num_beams=1, no_repeat_ngram_size=2)

generated_text = tokenizer.decode(output[0], skip_special_tokens=True)

print(generated_text)
```

Comment (reddyn12) [reply test/models/test_mamba.py:19]: If I take the trailing space off of the prompt, it does it in one line.

**Comment by geohot** [comment]
Once CLANG issue is fixed, I'll merge this and pay out bounty. I don't think we should run the model in CI, can just skip that test. It adds ~1 minute.

**Comment by reddyn12** [comment]
I put in the .realize between softmax and argmax for the time being and skipped the test in CI, should I add a benchmark run like with gpt?

**Comment by geohot** [comment]
"I put in the .realize between softmax and argmax' don't hide bugs!

**Comment by geohot** [comment]
Found a bug in testing:

```
diane@blak:~/tinygrad$ GPU=1 python3 examples/mamba.py --prompt "Hello."
ram used:  1.49 GB, lm_head.weight                                    : 100%|█████████████████████████████████████████████████████████████████████████████| 483/483 [00:00<00:00, 5055.32it/s]
loaded weights in 102.82 ms, 1.69 GB loaded at 16.46 GB/s
Speed Gen:   0%|                                                                                                                                                       | 0/10 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/Users/diane/tinygrad/examples/mamba.py", line 466, in <module>
    tinyoutput = generate(model, tokenizer, prompt, n_tokens_to_gen=num_toks)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 429, in generate
    logits = model(Tensor([tks[start_pos:]]), inference_params, start_pos, jit=False)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 390, in __call__
    return self.forward(input_ids, inference_params, num_last_tokens)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 379, in forward
    hidden_states = self.backbone(input_ids, inference_params=inference_params)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 356, in __call__
    hidden_states, residual = layer(
                              ^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 332, in __call__
    hidden_states = self.mixer(hidden_states, inference_params=inference_params)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/examples/mamba.py", line 224, in __call__
    conv_state.assign(x[:, :, -self.d_conv :])  # Update state (B D W)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/diane/tinygrad/tinygrad/tensor.py", line 146, in assign
    assert self.shape == x.shape, f"assign shape mismatch {self.shape} != {x.shape}"
           ^^^^^^^^^^^^^^^^^^^^^
AssertionError: assign shape mismatch (1, 2048, 4) != (1, 2048, 2)
```

**Comment by reddyn12** [comment]
I fixed the small-prompt bug. I'm looking over #3600 and I can't build on what @chenyuxyz discovered.

**Comment by chenyuxyz** [comment]
#3927 should fix the CLANG bug, can you test it (might need `DISABLE_COMPILER_CACHE=1` or clear the cache)?

**Comment by reddyn12** [comment]
The CLANG bug is fixed. However the .copyin() method in ops_cuda was changed from cuda.cuMemcpyHtoD_v2 to cuda.cuMemcpyHtoDAsync_v2 which breaks the load_state_dict() in nn.state. Can you confirm this breaks on multi cuda gpu, so we can switch back the method. If not, there has to be a single vs multi gpu if statement right?

**Comment by geohot** [comment]
Congrats, bounty is yours! Send either a PayPal e-mail or USDC ethereum address to claim, george@tinygrad.org

**Pull request merged**